director/origin_api_test.go

```diff
@@ -188,3 +193,182 @@ func TestNamespaceKeysCacheEviction(t *testing.T) {
 		}
 	})
 }
+
+// TestNamespaceKeysCacheTTLExpiration tests that the namespaceKeys cache
+// properly expires keys after the TTL period and that WithDisableTouchOnHit
+// prevents TTL refresh on cache access. This validates the fix for the bug
+// where stale keys would remain cached indefinitely if accessed frequently.
+func TestNamespaceKeysCacheTTLExpiration(t *testing.T) {
+	ctx, cancel, egrp := test_utils.TestContext(context.Background(), t)
+	defer func() { require.NoError(t, egrp.Wait()) }()
+	defer cancel()
+
+	server_utils.ResetTestState()
+
+	// Use a shorter TTL for testing (2 seconds instead of 15 minutes)
+	originalTTL := param.Director_AdvertisementTTL.GetDuration()
+	defer viper.Set(param.Director_AdvertisementTTL.GetName(), originalTTL)
+	viper.Set(param.Director_AdvertisementTTL.GetName(), 2*time.Second)
+
+	// Create test keys
+	oldKeyRaw, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	require.NoError(t, err, "Failed to generate old private key")
+	oldKey, err := jwk.FromRaw(oldKeyRaw)
+	require.NoError(t, err, "Failed to convert old key to JWK")
+	err = jwk.AssignKeyID(oldKey)
+	require.NoError(t, err)
+	err = oldKey.Set(jwk.AlgorithmKey, jwa.ES256)
+	require.NoError(t, err)
+
+	oldPublicKey, err := oldKey.PublicKey()
+	require.NoError(t, err, "Failed to create old public key")
+
+	newKeyRaw, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	require.NoError(t, err, "Failed to generate new private key")
+	newKey, err := jwk.FromRaw(newKeyRaw)
+	require.NoError(t, err, "Failed to convert new key to JWK")
+	err = jwk.AssignKeyID(newKey)
+	require.NoError(t, err)
+	err = newKey.Set(jwk.AlgorithmKey, jwa.ES256)
+	require.NoError(t, err)
+
+	newPublicKey, err := newKey.PublicKey()
+	require.NoError(t, err, "Failed to create new public key")
```

Comment:
I think it'd be preferable to use an existing function to generate the keys -- see `config/init_server_creds.go` and see if you can use one of the key generation functions there. Similarly, `test_utils/utils.go` provides `GenerateJWK()` and `GenerateJWKS()` that might be useful.

The first sign here that a helper function probably exists is that creating keys is a fairly common thing for us to do, and it's too easy to forget the bit about assigning a key ID (nice job remembering that, btw!). Whenever there's something common that's also nuanced/tricky to get right, there's probably a helper function somewhere. If there isn't, it's one of those situations where I wouldn't hesitate to add one and tell the rest of the dev team about it.